chain/chain/src/chain.rs

```diff
@@ -2765,13 +2770,15 @@ impl<'a> ChainUpdate<'a> {
         })
     }
 
+    // Applies chunks in separate threads. Returns callback that would actually wait till all
+    // chunks are applied
     fn apply_chunks(
         &mut self,
         me: &Option<AccountId>,
         block: &Block,
         prev_block: &Block,
         mode: ApplyChunksMode,
-    ) -> Result<(), Error> {
+    ) -> Result<impl FnOnce(&mut ChainUpdate) -> Result<(), Error>, Error> {
```

Comment:
This function is now very hard to read. The core issue is that it now scores high on three orthogonal complexity axis:

* it is just complex code (has many parameters, complex control flow)
* separately, it manages concurrency -- it spawns a concurrent thread of execution, which is not encapsulated to the function, but is returned to the caller as well. 
* separately, it involved complicated high-order language machinery -- rather than just computing stuff, it returns the closure to compute stuff later. 

I don't see a direct way to just make stuff simpler -- all three facets feel essential to how the code is written at the moment. But I wonder if we can restructure the whole computation such that we don't need this complexity. 

I think the original code was structured roughly like this:

```rust
fn apply_chunks(args) {
  for {
    let work = prepare_work();
    let result = do_work(work);
    apply_result(result)
  }
}

fn process_block() {
  if cond1 {
    apply_chunks(args1)
  }
  if cond2 {
    apply_chunks(args2)
  }
}
```

The structure in this PR looks like this:


```rust
fn apply_chunks(args) -> impl Fn() {
  let mut threads = Vec::new();
  for {
    let work = prepare_work();
    threads.push(spawn(|| do_work(work));
  }
  return || {
    for t in threads {
      let result = t.join();
      apply_result(result);
    }
  }
}

fn process_block() {
  let mut callbacks = Vec::new();
  if cond1 {
    callbacks.push(apply_chunks(args1))
  }
  if cond2 {
    callbacks.push(apply_chunks(args2))
  }
  for callback in callbacks {
    callback()
  }
}
```

I wonder if we can do some loop unswitching here, and express this as a parallel map? Something along these lines:

```rust
fn apply_chunks(args) -> Vec<Work> {
  let mut res = Vec::new()
  for {
    let work = prepare_work();
    res.push(work);
  }
  res
}

fn process_block() {
  let mut work: Vec<Work> = Vec::new();
  if cond1 {
    work.extend(apply_chunks(args1));
  }
  if cond2 {
    work.extend(apply_chunks(args2));
  }
  work
   .par_iter() // using rayon, declaratively process all work in parallel
   .map(|work| do_work(work))
   .for_each(|result| apply_result(result))
}
```

To clarify, I have a very superficial understanding of what's going on here, so I might be missing something obvious. But I do suggest trying to cast the problem as "First, prepare the work which can be done independently, then run each bit of work in parallel, then apply results".